### PR TITLE
Live: Plugin GRPC server prototype

### PR DESCRIPTION
**Pull request by FZambia**
**What this PR does / why we need it**:

Originally we need a simple way to publish data from a plugin to Grafana Live streams. We already can do this over a unidirectional GRPC stream initiated by Grafana Core - but for completeness we also need a way to publish into channels without using uni streams just using some generic endpoint:
* this can simplify a use case when managing unidirectional streams is too heavy/unjustified
* will allow maintaining publication stream history on Grafana Core side even if there are no active subscribers
* will allow custom publication processing even if there are no subscribers (save to Cloud?).

Since backend plugins already GRPC servers it seems reasonable to make Grafana a GRPC server also to open a reverse direction.

This is originally started here: https://github.com/grafana/grafana/pull/32722 - in that first iteration we aimed to make a generic GRPC server that can also be used not only by plugins. Now we start small - targeting only plugins. Though I think we should consider that this can also be used with user context in the future (not only within a plugin context).

### The good

* GRPC server protected by `plugin_api_server` feature flag, runs on unix and TCP sockets (TCP required for Windows)
* On Grafana start we generate JWT authentication keys for plugins (RSA256). Each time Grafana restarted a new private RSA key generated
* We pass auth key to a plugin over env variables
* GRPC client can automatically get authentication token from plugin environment, so for plugin API looks like shown here: https://github.com/grafana/grafana-plugin-sdk-go/pull/306
* Protobuf definitions for a server added as part of sdk pull request: https://github.com/grafana/grafana-plugin-sdk-go/pull/306 - since the initial goal is calling this server from plugins

Here is an example payload of token generated here for a plugin:

```
{
  "identity": "plugin",
  "iss": "grafana",
  "plugin": {
    "orgId": 1,
    "pluginId": "grafana-signal-datasource"
  },
  "sub": "plugin:1-grafana-signal-datasource"
}
```

### The bad

The bad thing is orgID handling, at this moment we hard-code orgID 1 in token passed to plugin. The solution to this could be  running separate datasource process per orgID, so each instance has its own well known identity. This is not that simple to refactor though and requires additional research.

### The ugly

* No tests at all yet
* Generating fake SignedInUser to get datasource - though DataSource Cache only uses orgID internally - so seems reasonable to refactor ds cache arguments?

### CODE REVIEW

**Comment by FZambia** [comment]
Updated with method to issue org ID token.

File: pkg/plugins/apiserver/server.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+func init() {
+	registry.RegisterServiceWithPriority(&GRPCAPIServer{}, registry.High)
+}
+
+// GRPCAPIServer ...
+type GRPCAPIServer struct {
+	Cfg             *setting.Cfg             `inject:""`
+	GrafanaLive     *live.GrafanaLive        `inject:""`
+	DatasourceCache datasources.CacheService `inject:""`
+	JWT             *authtoken.JWT           `inject:""`
```
Comment (ryantxu) [inline pkg/plugins/apiserver/server.go:42]: @malcolmholmes -- I got this much closer to compiling again, but the inject needs to be replaced with wire!

Comment (FZambia) [reply pkg/plugins/apiserver/server.go]: Just did this - all in wire now.

File: pkg/plugins/backendplugin/manager/manager.go [hunk 2]
```diff
@@ -66,6 +71,24 @@ func (m *Manager) Register(pluginID string, factory backendplugin.PluginFactoryF
...
+			grpcAddress = fmt.Sprintf("unix://%s", m.Cfg.GRPCServerAddress)
+		}
+		hostEnv = append(hostEnv, fmt.Sprintf("GF_GRPC_API_ADDRESS=%s", grpcAddress))
+
+		token, err := m.JWTService.IssuePluginToken(pluginID, 0)
+		if err != nil {
+			return err
+		}
+		m.logger.Warn("plugin token", "token", token)
+		hostEnv = append(hostEnv, fmt.Sprintf("GF_GRPC_API_TOKEN=%s", token))
```
Comment (ryantxu) [inline pkg/plugins/backendplugin/manager/manager.go:86]: :tada: plugins get started with a key to call the grafana server :tada:

File: pkg/plugins/apiserver/server.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+	JWT *authtoken.JWT,
+) (*GRPCAPIServer, error) {
+	s := &GRPCAPIServer{
+		Cfg:             cfg,
+		GrafanaLive:     GrafanaLive,
+		DatasourceCache: DatasourceCache,
+		JWT:             JWT,
+	}
+
+	logger.Info("Plugin GRPC API server initialization", "address", s.Cfg.GRPCServerAddress, "network", s.Cfg.GRPCServerNetwork, "tls", s.Cfg.GRPCServerUseTLS)
```
Comment (ryantxu) [inline pkg/plugins/apiserver/server.go:45]: This is a great start at a GRPC server in grafana -- however it is written so it is scoped for just plugins.  Do you think it makes sense so it is scoped at the root, but still accessible from plugins?

**Comment by FZambia** [comment]
Actual

**Pull request closed without merging**